MarsEnv.java

```diff
@@ -147,18 +162,120 @@ void nextSlot() throws Exception {
         }
 
         void moveTowards(int id, int x, int y) throws Exception {
-            Location r1 = getAgPos(id);
-            if (r1.x < x)
-                r1.x++;
-            else if (r1.x > x)
-                r1.x--;
-            else if (r1.y < y)
-                r1.y++;
-            else if (r1.y > y)
-                r1.y--;
-            setAgPos(id, r1);
-            setAgPos(1, getAgPos(1)); // just to draw it in the view
+            Location position = getAgPos(id);
+            if (position.x < x)
+                moveRight(id);
+            else if (position.x > x)
+                moveLeft(id);
+            else if (position.y < y)
+                moveDown(id);
+            else if (position.y > y)
+                moveUp(id);
         }
+		
+		/**
+		 * Implementation of the moveLeft action
+		 */
+		void moveLeft(int id) throws Exception {
+			Location position = getAgPos(id);
+			int x = position.x - 1;
+			int y = position.y;
+			move(id, x, y);
+		}
+		
+		/**
+		 * Implementation of the moveRight action
+		 */
+		void moveRight(int id) throws Exception {
+			Location position = getAgPos(id);
+			int x = position.x + 1;
+			int y = position.y;
+			move(id, x, y);
+		}
+
+		/**
+		 * Implementation of the moveUp action
+		 */
+		void moveUp(int id) throws Exception {
+			Location position = getAgPos(id);
+			int x = position.x;
+			int y = position.y - 1;
+			move(id, x, y);
+		}
+
+		/**
+		 * Implementation of the moveDown action
+		 */
+		void moveDown(int id) throws Exception {
+			Location position = getAgPos(id);
+			int x = position.x;
+			int y = position.y + 1;
+			move(id, x, y);
```

Comment:
```suggestion
			move(id, position.x, position.y+1); 
``` instead of the 3-line thing... but that's just aesthetic.

Reply:
This will appear in several places. Are you OK if I leave this for now at this revision and then change this on another pull request?

Reply:
sure, it works